### PR TITLE
Capture TurboBadger hashtable fix and more #1205

### PR DESCRIPTION
**Pull request by JimMarlowe**
Upgraded turbobadger some, it includes commits/prs:
ee05368  API for multitouch
898cc72  Added some unit tests & fixed bug in TBHashTable. 
TBWindow should have a virtual destructor #102
Fix TBImage::IsEmpty #93

I had to do some additional fixes for ee05368, I saw during compile that InvokePointerDown() and InvokePointerUp() were expecting to return bool, but they were still returning void, so I fixed them to the spirit in which they were intended.

Tested on Android as best as I could, see #1205 for details.

### CODE REVIEW

**Comment by JoshEngebretson** [comment]
Excellent, lots of good stuff pulled over 👍 I'll give this a once over and do some UI testing tomorrow on it too before landing. Thanks!

File: Source/ThirdParty/TurboBadger/tb_widgets.cpp [hunk 1]
```diff
@@ -1593,6 +1635,98 @@ void TBWidget::HandlePanningOnMove(int x, int y)
...
+        {
+            TBWidgetEvent ev(EVENT_TYPE_TOUCH_CANCEL, 0, 0, true);
+            ev.ref_id = id;
+            ti->captured_widget->InvokeEvent(ev);
+        }
+        DeleteTouchInfo(id);
+    }
+}
+
+
```
Comment (JoshEngebretson) [inline Source/ThirdParty/TurboBadger/tb_widgets.cpp]: Wow, this touch stuff is going to come in handy on mobile controls! 👍

**Comment by Type1J** [comment]
Touch is good for some laptops (Windows/Linux) as well. I wonder if a GeoSynth-like example with a dynamically generated saw, square or sine wave would be easy with Atomic after this PR.

**Comment by JimMarlowe** [comment]
I can reproduce it, will fix.

**Comment by JoshEngebretson** [comment]
@Type1J The new touch stuff needs to be wrapped in the UI subsystem and I think there are some TB cleanups (removals) that need to happen from some initial multitouch support added to Atomic back in 2015

**Comment by Type1J** [comment]
I assumed that we'd go through a clean up period. It's unfortunate that Atomic has some duplicated work, but it's good to see it on a path to consolidation. :smile:

**Comment by JimMarlowe** [comment]
I found the problem, I inserted the new touch events into the middle of the TB event enum, since that's where TB put them. Not knowing there is a parallel array for the UI, which is initialized with numbers, not the TB enum values, and this caused some odd behaviors, more would have popped up. I moved the new enums to the end of both lists, and all is well.

**Comment by JoshEngebretson** [comment]
I had to switch from the tb enums to use raw numbers for C# constants binding purposes, definitely preferred the matching enum values.  Probably worth a big WARNING comment in the TB enum, that it must be synced with the enum in UI subsystem... and  I think I have used up my allocated use of the keyword ```enum``` in this comment :)   I'll do some further testing before landing, thanks!

**Comment by JoshEngebretson** [comment]
Awesome, tests out, landing ✈️ , thanks! 🐫 ☀️ 😄

**Pull request merged**